Plugins: don't leave zombies

rustyrussell opened the pull request:
@jsarenik commented that he saw zombie plugins, and I reproduced that here.  This handles SIGCHLD in a way which tries not to interfere with the synchronous handling of waitpid() in the subd code.

Fixes #4381

Changelog-None

cdecker commented on lightningd/lightningd.c:
```diff
@@ -700,22 +700,79 @@ static void on_sigterm(int _ UNUSED)
+	 * there are no more children.  But glibc's overzealous use of
+	 * __attribute__((warn_unused_result)) means we have to
+	 * "catch" the return value. */
+        if (write(sigchld_wfd, "", 1));
```
`clang` complains that this is an empty `if`-statement. Likely just needs to be marked somehow.

cdecker commented on lightningd/plugin.c:
```diff
@@ -630,7 +648,8 @@ static struct io_plan *plugin_write_json(struct io_conn *conn,
 static void plugin_conn_finish(struct io_conn *conn, struct plugin *plugin)
 {
-	plugin_kill(plugin, "Plugin exited before completing handshake.");
+	const char *msg = tal_fmt(tmpctx, "died %s", state_desc(plugin));
```
Very much prefer the `exited` verbiage, since users usually get confused by terms like crash or dying, since they assume it's a fault.

rustyrussell commented:
OK, changed wording and hopefully make clang happier.

I also added better logging in general (logging at debug for normal expected plugin terminations).

cdecker commented:
Not up to date with signals, but somehow the signal handler is getting unexpected signals:

```
2021-02-11T06:34:59.1053093Z lightningd: FATAL SIGNAL 6 (version ea82286-modded)
2021-02-11T06:34:59.1053713Z 0x558a1a626f9b send_backtrace
2021-02-11T06:34:59.1054185Z 	common/daemon.c:39
2021-02-11T06:34:59.1054653Z 0x558a1a627045 crashdump
2021-02-11T06:34:59.1055099Z 	common/daemon.c:52
2021-02-11T06:34:59.1055531Z 0x7f38d71be20f ???
2021-02-11T06:34:59.1055894Z 	???:0
2021-02-11T06:34:59.1056249Z 0x7f38d71be18b ???
2021-02-11T06:34:59.1056620Z 	???:0
2021-02-11T06:34:59.1056969Z 0x7f38d719d858 ???
2021-02-11T06:34:59.1057336Z 	???:0
2021-02-11T06:34:59.1057681Z 0x7f38d719d728 ???
2021-02-11T06:34:59.1058030Z 	???:0
2021-02-11T06:34:59.1058401Z 0x7f38d71aef35 ???
2021-02-11T06:34:59.1058764Z 	???:0
2021-02-11T06:34:59.1059170Z 0x558a1a5cbe24 on_sigchild
2021-02-11T06:34:59.1059685Z 	lightningd/lightningd.c:718
2021-02-11T06:34:59.1060177Z 0x7f38d71be20f ???
2021-02-11T06:34:59.1060532Z 	???:0
2021-02-11T06:34:59.1060952Z 0x558a1a690a23 set_destroying_bit
2021-02-11T06:34:59.1061448Z 	ccan/ccan/tal/tal.c:103
2021-02-11T06:34:59.1061887Z 0x558a1a6912d8 del_tree
2021-02-11T06:34:59.1062336Z 	ccan/ccan/tal/tal.c:399
2021-02-11T06:34:59.1063039Z 0x558a1a691342 del_tree
2021-02-11T06:34:59.1063483Z 	ccan/ccan/tal/tal.c:412
2021-02-11T06:34:59.1064004Z 0x558a1a691342 del_tree
2021-02-11T06:34:59.1064438Z 	ccan/ccan/tal/tal.c:412
2021-02-11T06:34:59.1064895Z 0x558a1a691342 del_tree
2021-02-11T06:34:59.1065324Z 	ccan/ccan/tal/tal.c:412
2021-02-11T06:34:59.1065770Z 0x558a1a69168c tal_free
2021-02-11T06:34:59.1066198Z 	ccan/ccan/tal/tal.c:486
2021-02-11T06:34:59.1066633Z 0x558a1a5cc864 main
2021-02-11T06:34:59.1067116Z 	lightningd/lightningd.c:1125
2021-02-11T06:34:59.1067588Z 0x7f38d719f0b2 ???
2021-02-11T06:34:59.1067966Z 	???:0
2021-02-11T06:34:59.1068362Z 0x558a1a5af2ed ???
2021-02-11T06:34:59.1068750Z 	???:0
2021-02-11T06:34:59.1069168Z 0xffffffffffffffff ???
2021-02-11T06:34:59.1069593Z 	???:0
```

rustyrussell commented:
> Not up to date with signals, but somehow the signal handler is getting unexpected signals:
> 
> ```
> 2021-02-11T06:34:59.1051991Z lightningd: lightningd/lightningd.c:718: on_sigchild: Assertion `errno == EAGAIN || errno == EWOULDBLOCK' failed.

Ah!  Yes, it was getting EPIPE, since we closed the other end of the pipe when we freed ld.  Fix is simple, to deactivate signal handler before that.  See fixup.

jsarenik commented:
tACK 7ebef8f57c7fd425a11aa281e54ca4cfc75c094a (which is a no-op rebase of 5dd30b45f755d352e364c98733f63822b8704d36 to current master dc38e65ed3def291688fa434233a3184ac7fd9a3)

Everything works as expected. Thank you @rustyrussell !

cdecker commented:
Rebased and squashed

ACK https://github.com/ElementsProject/lightning/pull/4382/commits/53d7188c79b18838e245354846e9a9240ad03b56

jsarenik commented:
ACK 53d7188c79b18838e245354846e9a9240ad03b56
